sphinxsearch/backend/sphinx/compiler.py

```diff
@@ -96,12 +96,10 @@ def as_sql(self, with_limits=True, with_col_aliases=False, subquery=False):
         # of SQL expressions in sphinxsearch.
         try:
             self._add_where_result(query, where)
-        except EmptyResultSet:
+        finally:
```

Comment:
Это требует отдельного пояснения.

Reply:
В данный момент попытка выбрать срез queryset-a с индексом [0:0] и [:0] приводит к выдаче результатов без учёта фильтров ноды django where и индексов вовсе. 
Стандартный порядок обработки Django таких индексов такой:

1. При указании индекса [:0] или [0:0] django добавляет в childrens ноды WhereNode ноду NothingNode.
2. При компиляции query одной из первых компилируется нода where. NothingNode при компиляции вызывает exception EmptyResultSet, в следствие чего where, limits и прочее не компилируются.
3. Django обрабатывая exception EmptyResultSet возвращая пустой кверисет.

В нашем случае мы гасили EmptyResultSet, в следствие чего where, limits и прочее не формировались, но и нужный exception до django не доходил и она просто обрабатывала запрос как будто всё в порядке.

А гасили мы его для потому, что нам в любом случае надо было выполнить родной `SQLCompiler.as_sql`  для вызова pre_sql_setup и пр. 
Я добавил вызов `SQLCompiler.as_sql` в final чтобы он выполнялся, но exception проходил дальше. 
Без этого не проходит добавленный мной тест testLenOfAnotherEmptySet.

PS. При этом, если без этой правки сначала сформировать queryset, вызвав len(qs) или list(qs), а потом уже получить его пустой срез, то результат был действительно пустым и выглядел корректно, поэтому эта ошибка оставалась незамеченной. Она стала мешать после обновления на нашем проекте django до 2.2. Стандартный paginator начал возвращать срез [0:0] если проверка qs.count() показывала отсутствие результатов.

Reply:
Мне кажется, основная проблема несколько выше:
```
query = self.query
        # replacing WHERE node with sphinx-aware node implementation
        where, query.where = query.where, sqls.SphinxWhereNode()
        # Adding MATCH() node to WHERE node if needed
        self._add_match_extra(query)
        # moving where conditions to SELECT clause because of better support
        # of SQL expressions in sphinxsearch.
        try:
            self._add_where_result(query, where)
        except EmptyResultSet:
            # Where node compiled to always-false condition, but we still need
            # to call pre_sql_setup() and other methods by super().as_sql
            pass
```

1. Сначала мы меняем состояние `query`, сбросив ему `where` на пустую ноду.
2. Затем мы в эту пустую ноду переносим условия из `match`
3. Псле этого мы компилируем исходный `where`, для того чтобы засунуть в `SELECT ... as __where_result` и получаем ошибку
4. Тут по задумке мы должны вернуть query к исходному состоянию, но походу query.where остается в некоторых случаях пустым и в результате вызов `super().as_sql` не приводит к той же ошибке

Мне кажется, надо переделать работу всех указанных в сниппете строк:
```
        try:
            where = sqls.SphinxWhereNode()
            self._add_match_extra(query, where)
            self._add_where_result(query, where, query.where)
            query.where = where
        except EmptyResultSet:
            pass
```
Идея в том, чтобы до последнего не модифицировать вообще никак объект query, и только после того, как получим успешно скомпилированный where, проставить его в query.

Твой вариант как минимум пытается выполнять as_sql и в любых других случаях, но в целом проблему ошибки и отката на mysql-версию не решает.

Reply:
Давай уточним что я правильно понял что ты хочешь.

Ты предлагаешь скомпилировать where не модифицируя query и в случае, если это удалось, добавить туда условия из match и в _add_where_result убрать содержимое полученной where и перенести его в аннотацию?

И главное, что ты хочешь сделать если where не удаётся скомпилировать из-за исключения EmptyResultSet? В except ты по прежнему гасишь исключение. Логично туда вставить super().as_sql, который вызовет pre_sql_setup, который, в свою очередь, заполнит как минимум `compiler.klass_info`, без чего Model.Iterable не сможет проитерировать queryset даже если он будет пустой. А затем вызовет исключение при повторной компиляции ноды where.

Соответственно, если я правильно понял, ты не хочешь делать лишние действия по конвертации where в формат сфинкс если уже понятно что кверисет пуст?

Если я изложил верно, то это предложение мне не понятно:
> Твой вариант как минимум пытается выполнять as_sql и в любых других случаях, но в целом проблему ошибки и отката на mysql-версию не решает.

Проблема ошибки, как я её понимаю, это подавление эксепшена EmptyResultSet, который в:
```
def execute_sql(...
...
        try:
            sql, params = self.as_sql()
            if not sql:
                raise EmptyResultSet
        except EmptyResultSet:
            if result_type == MULTI:
                return iter([])
            else:
                return
```
возвращает пустые результаты в `ModelIterable.__iter__`. По идее мой вариант как раз её решает. Что значит откат на `mysql-версию`?

Reply:
Ответы на знаки вопроса: да, да, да; откат на mysql-версию опишу чуть подробнее.

Каким образом вообще появляется в Where EmptyResultSet? Он там появляется в тот момент, когда по каким-то эвристикам получается условие, которое всегда False.
```qs.filter(field__isnull=True, field__isnull=False)```

Компиляция Match в теории к этой ошибке в принципе привести не может, потому что таких эвристик там не предусмотрено. Поэтому если мы в нашем модифицированном бекенде, основанном на `django.db.backends.mysql` словили эту ошибку, это означает, что и mysql-backend ее тоже вбросит. Именно поэтому EmptyResultSet в исходном варианте просто пропускался, полагаясь на то, что вызов super().as_sql() всегда приведет к тому же варианту.

Теперь насчет отличий твоего решения и предложенного мной.

* Исходное решение имеет побочные эффекты в виде замены query.where на пустое условие
* Твое решение тоже имеет тот же побочный эффект
* Мое решение от этого эффекта избавлено, т.к. сначала ловит нужное исключение, а ктолько потом изменяет query.where

Собственно из-за того, весь сыр бор как раз из-за этого побочного эффекта, так что оставлять его в живых - бессмысленно и вообще опасно.